src/app/util/ember-compatibility-functions.cpp

```diff
@@ -210,222 +221,254 @@ bool ServerClusterCommandExists(const ConcreteCommandPath & aCommandPath)
     return emberAfContainsServer(aCommandPath.mEndpointId, aCommandPath.mClusterId);
 }
 
-CHIP_ERROR ReadSingleClusterData(FabricIndex aAccessingFabricIndex, const ConcreteAttributePath & aPath,
+CHIP_ERROR ReadSingleClusterData(const SubjectDescriptor & aSubjectDescriptor, const ConcreteAttributePath & aPath,
                                  AttributeReportIB::Builder & aAttributeReport)
 {
     ChipLogDetail(DataManagement,
                   "Reading attribute: Cluster=" ChipLogFormatMEI " Endpoint=%" PRIx16 " AttributeId=" ChipLogFormatMEI,
                   ChipLogValueMEI(aPath.mClusterId), aPath.mEndpointId, ChipLogValueMEI(aPath.mAttributeId));
-    AttributeDataIB::Builder attributeDataIBBuilder;
-    AttributePathIB::Builder attributePathIBBuilder;
-    AttributeStatusIB::Builder attributeStatusIBBuilder;
-    TLV::TLVWriter * writer = nullptr;
+
     TLV::TLVWriter backup;
     aAttributeReport.Checkpoint(backup);
 
-    attributeDataIBBuilder = aAttributeReport.CreateAttributeData();
-    attributePathIBBuilder = attributeDataIBBuilder.CreatePath();
+    AttributeDataIB::Builder attributeDataIBBuilder = aAttributeReport.CreateAttributeData();
+    AttributePathIB::Builder attributePathIBBuilder = attributeDataIBBuilder.CreatePath();
     attributePathIBBuilder.Endpoint(aPath.mEndpointId)
         .Cluster(aPath.mClusterId)
         .Attribute(aPath.mAttributeId)
         .EndOfAttributePathIB();
     ReturnErrorOnFailure(attributePathIBBuilder.GetError());
 
-    AttributeAccessInterface * attrOverride = findAttributeAccessOverride(aPath.mEndpointId, aPath.mClusterId);
-    if (attrOverride != nullptr)
+    Protocols::InteractionModel::Status status = Protocols::InteractionModel::Status::Failure;
+
     {
-        // TODO: We should probably clone the writer and convert failures here
-        // into status responses, unless our caller already does that.
-        writer = attributeDataIBBuilder.GetWriter();
+        TLV::TLVWriter * writer = attributeDataIBBuilder.GetWriter();
```

Comment:
Thank you for fixing this, was annoying me for a while...

Reply:
I didn't do anything special though to clone the writer, I don't think. I just reordered all the calls so they occur when needed, and ExitNow (goto) doesn't jump across object lifetimes. It should overall make more sense in the flow of the function, though. I think more work on this front is still required.

Reply:
I did set up the functions so that it's easy to return a status response: just set the IM status, and ExitNow(), then the exit: labelled code checks it and handles it. I needed to do that for access control to easily return DENIED.

Reply:
BTW the TODO lives on, it just moved to line 256, which is only a few away, but very far in this diff view on github.